source/tests/system/visa_driver_api_tests.cpp

```diff
@@ -153,6 +153,149 @@ TEST_F(VisaDriverApiTest, SetNonTcpipAttribute_ReturnsAttributeError)
+    server_address.sin_port = htons(std::stoi(portNumber));
+    EXPECT_EQ(0, bind(server_fd_, (struct sockaddr*)&server_address, sizeof(server_address)));
+
+    server_thread_ = std::thread(&VisaDriverLoopbackTest::run_server, this, server_fd_);
+  }
+
+  void run_server(SOCKET server_fd) {
+    listen(server_fd, 1);
+    SOCKET client_fd = accept(server_fd, NULL, NULL);
+    if (client_fd == -1) {
+      return;
+    }
+    std::thread([this, client_fd]() {
+      handle_client(client_fd);
+    }).detach();
+  }
+
+  void close_server_session()
+  {
+    server_thread_.join();
+  #ifdef _WIN32
+    closesocket(server_fd_);
+    WSACleanup();
+  #else
+    close(server_fd_);
+  #endif
+  }
+
+  void handle_client(SOCKET client_fd)
+  {
+    auto session = std::make_shared<TcpEchoSession>(client_fd);
+    session->start();
+  }
+
+  void write(const std::string& data)
+  {
+    auto response = client::write(GetStub(), GetNamedSession(), data);
+    EXPECT_EQ(VI_SUCCESS, response.status());
+    EXPECT_EQ(data.size(), response.return_count());
+  }
+
+  void read(uint32_t count, const std::string& expectedData, ViStatus expectedStatus)
+  {
+    auto response = client::read(GetStub(), GetNamedSession(), count);
+    EXPECT_EQ(expectedStatus, response.status());
+    EXPECT_EQ(expectedData.size(), response.return_count());
+    EXPECT_EQ(expectedData, response.buffer());
+  }
+
+  private:
+    std::string portNumber_;
+    std::string instrument_descriptor_;
+    SOCKET server_fd_;
+    std::thread server_thread_;
+};
+
+TEST_F(VisaDriverLoopbackTest, WriteAndRead_Matches)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData);
+  read(writeData.size(), writeData, VI_SUCCESS_MAX_CNT);
+}
+
+TEST_F(VisaDriverLoopbackTest, WriteTwice_ReadOnce_Matches)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData + "1 ");
+  write(writeData + "2");
+  std::string expectedReadData = writeData + "1 " + writeData + "2";
+  read(expectedReadData.size(), expectedReadData, VI_SUCCESS_MAX_CNT);
+}
+
+TEST_F(VisaDriverLoopbackTest, ReadMoreThanWritten_ReturnTimeout)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData);
+  read(writeData.size() + 1, writeData, VI_ERROR_TMO);
+}
+
+TEST_F(VisaDriverLoopbackTest, ReadLessThanWritten_ReturnSuccess)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData);
+  read(writeData.size() - 1, writeData.substr(0, writeData.size() - 1), VI_SUCCESS_MAX_CNT);
+}
+
+TEST_F(VisaDriverLoopbackTest, ReadZeroBytes_ReturnSuccess)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData);
+  read(0, "", VI_SUCCESS_MAX_CNT);
+}
+
+TEST_F(VisaDriverLoopbackTest, WriteOnce_ReadTwice_Matches)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData);
+  read(10, writeData.substr(0, 10), VI_SUCCESS_MAX_CNT);
+  read(writeData.size() - 10, writeData.substr(10, writeData.size()), VI_SUCCESS_MAX_CNT);
+}
+
```

Comment:
One test that might be useful is to set the attribute TERMCHAR_EN to true. Then write "Hello\nWorld\n" and do read(1024) twice. The first will return "Hello\n" and the second read will return "World\n" and both will be VI_SUCCESS_TERMCHAR.